### PR TITLE
Use symbol server when the pdb doesn't exist locally

### PR DESCRIPTION
**Pull request by calixteman**
To configure the symbol server:
- create a file ~/.dump_syms/config
- add the lines: 
SRV\*\~/symcache\*https://msdl.microsoft.com/download/symbols
SRV\*\~/symcache\*https://symbols.mozilla.org/

If the pdb cannot be found locally then in using the ids we've in the PE we can try to get the pdb on those servers

### CODE REVIEW

File: src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        .unwrap_or_else(|_| panic!("Cannot open file {} for writing", path.to_str().unwrap()));
+    let mut output = BufWriter::new(output);
+    output
+        .write_all(data)
+        .unwrap_or_else(|_| panic!("Cannot write file {}", path.to_str().unwrap()));
+
+    true
+}
+
+pub fn search_symbol_file(file_name: String, debug_id: &str) -> (Option<Vec<u8>>, String) {
```
Comment (gabrielesvelto) [inline src/cache.rs]: This function is too large and does a lot of different things, please split it out.

File: src/windows/pdb.rs [hunk 2]
```diff
@@ -305,12 +306,20 @@ impl PDBInfo<'_> {
...
+        let pe_debug_id = get_pe_debug_id(pe.as_ref());
+
+        if let Some(pe_debug_id) = pe_debug_id {
+            if debug_id != pe_debug_id {
+                return Err(From::from(format!(
+                    "Debug ids aren't the same: {} (pdb) != {} (pe)",
+                    debug_id, pe_debug_id
+                )));
+            }
+        }
```
Comment (gabrielesvelto) [inline src/windows/pdb.rs]: This change is unrelated to the rest, split it out in a separate commit.This change is unrelated to the rest, split it out in a separate commit. It would also be nice to have a test for it.

File: src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+fn copy_in_cache(path: Option<PathBuf>, data: &[u8]) -> bool {
+    if data.is_empty() || data.starts_with(b"Symbol Not Found") {
+        return false;
+    }
+
+    if path.is_none() {
+        return true;
+    }
+
+    let path = path.unwrap();
```
Comment (gabrielesvelto) [inline src/cache.rs]: It's more idiomatic to use `if let` here instead of checking if `path` is none. Especially since you also return true later.

**Comment by gabrielesvelto** [comment]
Also I'm curious: did we already have this functionality in `dump_syms`? I've never used it before so I don't know.

**Comment by calixteman** [comment]
> Also I'm curious: did we already have this functionality in `dump_syms`? I've never used it before so I don't know.

No it didn't exist.
I though it'll make our life easier.

**Comment by gabrielesvelto** [comment]
> 
> 
> > Also I'm curious: did we already have this functionality in `dump_syms`? I've never used it before so I don't know.
> 
> No it didn't exist.
> I though it'll make our life easier.
> Anyway thanks to that feature I found the bug about pdb containing only public symbols in testing with a ms pdb.

OK, sounds good. I just wonder if using a configuration file is the right way to go considering that most of our automation is stateless. Would it be complicated to use an optional command-line parameter instead?

**Comment by calixteman** [comment]
I think we can have both of them.
For automation, an option is better but for debug purpose it'll be easier to have it in a conf file.
